### ISSUE TITLE
Database name seems to be hardcoded

### ISSUE THREAD

**Issue by davidmehren** [hedgedoc/container#136]
@ErikMichelson and I noticed a very weird behavior:

1. Use the example `docker-compose.yml` with MariaDB
```yaml
version: '3'
services:
  database:
    image: mariadb:10
    environment:
      - MYSQL_USER=hedgedoc
      - MYSQL_PASSWORD=password
      - MYSQL_DATABASE=hedgedoc
      - MYSQL_ALLOW_EMPTY_PASSWORD=true
    volumes:
      - database:/var/lib/mysql
      - ./resources/utf8.cnf:/etc/mysql/conf.d/utf8.cnf
    networks:
      backend:
    restart: always
  app:
    image: quay.io/hedgedoc/hedgedoc:1.6.0
    environment:
      - CMD_DB_URL=mariadb://hedgedoc:password@database:3306/hedgedoc
    volumes:
      - uploads:/hedgedoc/public/uploads
    ports:
      - "3000:3000"
    networks:
      backend:
    restart: always
    depends_on:
      - database
networks:
  backend:
volumes:
  database:
  uploads:
```
2. Start the stack. Everything starts up normally.
3. Change the image to `quay.io/codimd/server:1.6.0`, `down` and `up` the stack.
4. HedgeDoc now fails to start:
```
app_1       | Sequelize CLI [Node: 12.19.0, CLI: 5.5.1, ORM: 5.21.4]
app_1       | 
app_1       | Parsed url mariadb://hedgedoc:*****@database:3306/hedgedoc
app_1       | warning: please use IANA standard timezone format ('Etc/GMT0')
app_1       | warning: please use IANA standard timezone format ('Etc/GMT0')
app_1       | No migrations were executed, database schema was already up to date.
database_1  | 2020-12-03  9:43:54 5 [Warning] Aborted connection 5 to db: 'hedgedoc' user: 'hedgedoc' host: '172.23.0.3' (Got an error reading communication packets)
app_1       | 2020-12-03T09:43:57.532Z warn: 	Neither 'domain' nor 'CMD_DOMAIN' is configured. This can cause issues with various components.
app_1       | Hint: Make sure 'protocolUseSSL' and 'urlAddPort' or 'CMD_PROTOCOL_USESSL' and 'CMD_URL_ADDPORT' are configured properly.
app_1       | 2020-12-03T09:43:57.534Z warn: 	Session secret not set. Using random generated one. Please set `sessionSecret` in your config.js file. All users will be logged out.
app_1       | 2020-12-03T09:43:57.535Z warn: 	PDF export was disabled for this release to mitigate a critical security issue. This feature will hopefully become available again in future releases.
app_1       | Thu, 03 Dec 2020 09:43:57 GMT hsts deprecated The "includeSubdomains" parameter is deprecated. Use "includeSubDomains" (with a capital D) instead. at app.js:85:18
app_1       | warning: please use IANA standard timezone format ('Etc/GMT0')
app_1       | warning: please use IANA standard timezone format ('Etc/GMT0')
app_1       | Unhandled rejection SequelizeDatabaseError: (conn=7, no: 1142, SQLState: 42000) SELECT command denied to user 'hedgedoc'@'172.23.0.3' for table 'Users'
app_1       | sql: SHOW INDEX FROM `Users` FROM `hackmd` - parameters:[]
app_1       |     at Query.formatError (/codimd/node_modules/sequelize/lib/dialects/mariadb/query.js:285:16)
app_1       |     at /codimd/node_modules/sequelize/lib/dialects/mariadb/query.js:68:22
app_1       |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
```
The two `FROM`s in `sql: SHOW INDEX FROM `Users` FROM `hackmd` - parameters:[]` look like Sequelize is for some reason trying to use a `hackmd` database. Note that the URL still uses `hedgedoc` and the `sequelize-cli` uses the correct database to check for migrations.

This problem seems to also occur the other way round: If you previously deployed the old `codimd/server` image and used a database named `hackmd` ([the default](https://github.com/hedgedoc/container/commit/298f793c5c59984979b6561261d6c418d7a003ef#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3L70) before 298f793c5c59984979b6561261d6c418d7a003ef) and now upgrade to a image under `hedgedoc/hedgedoc` without touching anything else, HedgeDoc fails to connect to the database (as it wants to use `hedgedoc`).

So it looks like we use a database called `hackmd` for images before 298f793c5c59984979b6561261d6c418d7a003ef and a database called `hedgedoc` for images after that, regardless of what is configured in `CMD_DB_URL`?

**Comment by davidmehren** [comment]
I started a `quay.io/codimd/server:1.6.0` container using the docker-compose config above. Expectedly, it crashed:
```
app_1       | Unhandled rejection SequelizeDatabaseError: (conn=7, no: 1142, SQLState: 42000) SELECT command denied to user 'hedgedoc'@'172.18.0.3' for table 'Users'
app_1       | sql: SHOW INDEX FROM `Users` FROM `hackmd` - parameters:[]
```

Using `docker exec -it container_app_1 /bin/bash` I entered the container, installed `nano` and edited the `config.json` to use `"database": "hackmd123"`. 
After a restart, the container crashed with:
```
app_1       | Unhandled rejection SequelizeDatabaseError: (conn=7, no: 1142, SQLState: 42000) SELECT command denied to user 'hedgedoc'@'172.18.0.3' for table 'Users'
app_1       | sql: SHOW INDEX FROM `Users` FROM `hackmd123` - parameters:[]
```

This clearly shows, that the database name is taken from the `config.json` and therefore hardcoded.

**Comment by davidmehren** [comment]
I can also reproduce this without docker:
- Start MariaDB, create a database `hedgedoc` and a user `hedgedoc` with password `hedgedoc`.
- Setup environment variables: `NODE_ENV=production` and `CMD_DB_URL=mariadb://hedgedoc:hedgedoc@127.0.0.1:3306/hedgedoc`.
- Use a `config.json` like this:
	```
	{
	  "production": {
	    "db": {
	      "username": "hedgedoc",
	      "password": "hedgedoc",
	      "database": "hedgedoc1",
	      "host": "localhost",
	      "port": "3306",
	      "dialect": "mariadb"
	    }
	  }
	}
	```
	Note that the database is called `hedgedoc1` here.
- Start HedgeDoc using `node app.js`
- HedgeDoc will crash on startup:
	```
	Unhandled rejection SequelizeDatabaseError: (conn=504, no: 1142, SQLState: 42000) SELECT command denied to user 'hedgedoc'@'localhost' for table 'Users'
	sql: SHOW INDEX FROM `Users` FROM `hedgedoc1` - parameters:[]
	    at Query.formatError (/home/david/Projekte/hedgedoc/legacy/node_modules/sequelize/lib/dialects/mariadb/query.js:285:16)
	    at /home/david/Projekte/hedgedoc/legacy/node_modules/sequelize/lib/dialects/mariadb/query.js:68:22
	    at processTicksAndRejections (node:internal/process/task_queues:93:5)
	```
- When using `mysql` as the db driver, the error message looks different:
	```
	Unhandled rejection SequelizeDatabaseError: SELECT command denied to user 'hedgedoc'@'localhost' for table 'Users'
	    at Query.formatError (/home/david/Projekte/hedgedoc/legacy/node_modules/sequelize/lib/dialects/mysql/query.js:244:16)
	    at Query.handler [as onResult] (/home/david/Projekte/hedgedoc/legacy/node_modules/sequelize/lib/dialects/mysql/query.js:51:23)
	    at Query.execute (/home/david/Projekte/hedgedoc/legacy/node_modules/mysql2/lib/commands/command.js:30:14)
	    at Connection.handlePacket (/home/david/Projekte/hedgedoc/legacy/node_modules/mysql2/lib/connection.js:425:32)
	    at PacketParser.onPacket (/home/david/Projekte/hedgedoc/legacy/node_modules/mysql2/lib/connection.js:75:12)
	    at PacketParser.executeStart (/home/david/Projekte/hedgedoc/legacy/node_modules/mysql2/lib/packet_parser.js:75:16)
	    at Socket.<anonymous> (/home/david/Projekte/hedgedoc/legacy/node_modules/mysql2/lib/connection.js:82:25)
	    at Socket.emit (node:events:376:20)
	    at addChunk (node:internal/streams/readable:305:12)
	    at readableAddChunk (node:internal/streams/readable:280:9)
	    at Socket.Readable.push (node:internal/streams/readable:219:10)
	    at TCP.onStreamRead (node:internal/stream_base_commons:192:23)
	```

**Comment by davidmehren** [comment]
This looks like a sequelize bug:
- At/around https://github.com/sequelize/sequelize/blob/v5/lib/sequelize.js#L198 the URL is parsed and the results are stored in the `config` object.
- The `options` object comes from the constructor and in our case contains config from `config.json`. It is saved to `this.options`.
- After https://github.com/sequelize/sequelize/blob/v5/lib/sequelize.js#L294 `this.config` contains the merged config from `config` and `this.options` (great variable naming, btw :smile:).
- The `sync` method at https://github.com/sequelize/sequelize/blob/v5/lib/sequelize.js#L796 uses `this.options` (instead of `this.config`) and therefore only uses the values from our `config.json`.

Next up, I'll open an issue over at sequelize.

**Comment by davidmehren** [comment]
I'm thinking about the best (or 'least hacky') workaround for this:
As the URL parsing/handling of sequelize is broken, we could parse the DB URL ourselves and pass the 'parts' to sequelize, using the normal constructor arguments.

**Comment by SISheogorath** [comment]
Did sequelize decided that this is "won't fix"?

Also if I understand it correctly its a problem that mainly affects the `sync` function? We could back port the migration change with umzug that I build for 2.0 before the rewrite and drop this in here in order to get rid of the sync call.

**Comment by davidmehren** [comment]
> Did sequelize decided that this is "won't fix"?

If have yet to open a bug for that. I'm looking for a workaround because the Sequelize maintainers currently don't have much time for the project, so I don't expect a quick fix.
As I'd like to release 1.7 soon, finding a workaround seems to be a more practical way.

**Comment by SISheogorath** [comment]
So writing an own URL parser is easier than fixing the one in sequelize?

**Comment by ErikMichelson** [comment]
My favorite solution for this is to backport the umzug code. I don't know it exactly but could imagine, that it would be the easiest possibility to fix this.
The bug should be opened at sequelize nevertheless.

**Comment by davidmehren** [comment]
After *even more* debugging, I found that I wasn't completely right:
Actually, `this.config` (and therefore the correct configuration) **is** used to connect to the database.
There is only one place where the database name from `this.options` is used: https://github.com/sequelize/sequelize/blob/v5/lib/dialects/mysql/query-generator.js#L303  
Why this query includes the database name is unclear. It was introduced in https://github.com/sequelize/sequelize/commit/e21579b802bcb4e89293af22b14c8d7a9a00fc98, but the commit message is not very helpful :disappointed:.

To summarize: The `mysql` and `mariadb` (which is based on `mysql`) dialects do not work when a different database name than the one on `config.json` is used.

I propose to remove the hardcoded `db.database` key from the `config.json`, which will lead to the query-generator not appending a second `FROM` as `options.database` will be undefined. I tested this fix with a locally built docker image.
